### PR TITLE
Add nanoflow behaviour

### PR DESCRIPTION
**Pull request by jseremba**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,19 +1,34 @@
...
-    "type": "git",
-    "url": "git+https://github.com/mendix/CameraWidgetForPhoneGap.git"
-  },
-  "author": "Mendix Technology bv",
   "license": "Apache-2.0",
-  "bugs": {
-    "url": "https://github.com/mendix/CameraWidgetForPhoneGap/issues"
-  },
-  "homepage": "https://github.com/mendix/CameraWidgetForPhoneGap#readme",
+  "author": "Mendix Technology bv",
```
Comment (McDoyen) [inline package.json]: Mendix BV

File: package.json [hunk 2]
```diff
@@ -1,19 +1,34 @@
 {
-  "name": "camera-widget-for-phonegap",
+  "name": "CameraWidgetForPhoneGap",
   "version": "3.2.0",
   "description": "Enable native camera functionality within your Mendix hybrid app",
-  "repository": {
```
Comment (McDoyen) [inline package.json]: Why did you remove `repository`

File: src/CameraWidgetForPhoneGap/CameraWidgetForPhoneGap.xml [hunk 3]
```diff
@@ -63,6 +63,12 @@
             <description>This mf is called when the object is saved. use this to commit the object when using autosave.</description> 
             <returnType type="Void"/>
         </property>
+        <property key="onChangeNanoFlow" type="nanoflow" required="false">
```
Comment (McDoyen) [inline src/CameraWidgetForPhoneGap/CameraWidgetForPhoneGap.xml]: `Nanoflow` is one word.
> Use onSaveNanoflow

File: src/CameraWidgetForPhoneGap/widget/CameraWidgetForPhoneGap.js [hunk 4]
```diff
@@ -20,6 +20,7 @@ require([
         targetHeight: 150,
         autoSaveEnabled: false,
         onchangemf: "",
+        onChangeNanoFlow: "",
```
Comment (McDoyen) [inline src/CameraWidgetForPhoneGap/widget/CameraWidgetForPhoneGap.js]: >onSaveNanoflow

**Review by McDoyen — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (jseremba) [reply package.json]: Changed

File: package.json (hunk 2 above)
Comment (jseremba) [reply package.json]: It wasnt originally there but i've added it.

File: src/CameraWidgetForPhoneGap/CameraWidgetForPhoneGap.xml (hunk 3 above)
Comment (jseremba) [reply src/CameraWidgetForPhoneGap/CameraWidgetForPhoneGap.xml]: Done

File: src/CameraWidgetForPhoneGap/widget/CameraWidgetForPhoneGap.js (hunk 4 above)
Comment (jseremba) [reply src/CameraWidgetForPhoneGap/widget/CameraWidgetForPhoneGap.js]: Changed

File: package.json [hunk 5]
```diff
@@ -1,19 +1,38 @@
 {
-  "name": "camera-widget-for-phonegap",
-  "version": "3.2.0",
+  "name": "CameraWidgetForPhoneGap",
+  "version": "3.3.0",
   "description": "Enable native camera functionality within your Mendix hybrid app",
+  "license": "Apache-2.0",
+  "author": "Mendix BV",
+  "private": true,
```
Comment (Andries-Smit) [inline package.json]: remove `"private": true,` its public

File: package.json [hunk 6]
```diff
@@ -1,19 +1,38 @@
...
   "repository": {
     "type": "git",
-    "url": "git+https://github.com/mendix/CameraWidgetForPhoneGap.git"
-  },
-  "author": "Mendix Technology bv",
-  "license": "Apache-2.0",
-  "bugs": {
-    "url": "https://github.com/mendix/CameraWidgetForPhoneGap/issues"
+    "url": "https://github.com/mendix/CameraWidgetForPhoneGap"
```
Comment (Andries-Smit) [inline package.json]: incorrect should be `"url": "https://github.com/mendix/CameraWidgetForPhoneGap/issues"`

File: package.json [hunk 7]
```diff
@@ -1,19 +1,38 @@
...
   },
-  "homepage": "https://github.com/mendix/CameraWidgetForPhoneGap#readme",
```
Comment (Andries-Smit) [inline package.json]: why remove this? was it incorrect?

File: src/CameraWidgetForPhoneGap/CameraWidgetForPhoneGap.xml [hunk 8]
```diff
@@ -63,6 +63,12 @@
             <description>This mf is called when the object is saved. use this to commit the object when using autosave.</description>
```
Comment (Andries-Smit) [inline src/CameraWidgetForPhoneGap/CameraWidgetForPhoneGap.xml]: change to microflow

File: src/CameraWidgetForPhoneGap/CameraWidgetForPhoneGap.xml [hunk 9]
```diff
@@ -63,6 +63,12 @@
             <description>This mf is called when the object is saved. use this to commit the object when using autosave.</description> 
             <returnType type="Void"/>
         </property>
+        <property key="onSaveNanoflow" type="nanoflow" required="false">
+            <caption>On save nanoflow</caption>
+            <category>Behaviour</category>
+            <description>This nano flow is called when the object is saved. use this to commit the object when using autosave</description>
```
Comment (Andries-Smit) [inline src/CameraWidgetForPhoneGap/CameraWidgetForPhoneGap.xml]: one word `nanoflow`

File: src/CameraWidgetForPhoneGap/widget/CameraWidgetForPhoneGap.js [hunk 10]
```diff
@@ -253,6 +254,17 @@ require([
...
             }
-        }
+
+            if (this.onSaveNanoflow && this.mxcontext) {
+                mx.data.callNanoflow({
+                    nanoflow: this.onSaveNanoflow,
+                    origin: this.mxform,
+                    context: this.mxcontext,
+                    error: function (error) {
+                        mx.ui.error(error.message);
```
Comment (Andries-Smit) [inline src/CameraWidgetForPhoneGap/widget/CameraWidgetForPhoneGap.js]: Add better description. `"Error executing nanoflow " + this.onSaveNanoflow + " : " + error.message`

**Review by Andries-Smit — CHANGES_REQUESTED** [review]

File: src/CameraWidgetForPhoneGap/widget/CameraWidgetForPhoneGap.js [hunk 11]
```diff
@@ -154,36 +154,20 @@ require([
...
-                        fileReader.onerror = function(e) {
-                            error(e.target.error);
-                        };
-
-                        fileReader.readAsArrayBuffer(blob);
-                    }, error);
+            window.resolveLocalFileSystemURL(this._imageUrl, function (fileEntry) {
+                fileEntry.file(function (blob) {
+                    var fileReader = new FileReader();
+                    fileReader.onload = function (e) {
```
Comment (jseremba) [inline src/CameraWidgetForPhoneGap/widget/CameraWidgetForPhoneGap.js]: remove short form `e` -> `event`

File: src/CameraWidgetForPhoneGap/widget/CameraWidgetForPhoneGap.js [hunk 12]
```diff
@@ -154,36 +154,20 @@ require([
...
+                        window.mx.data.saveDocument(guid, filename, {}, new Blob([e.target.result]), success, error);
+                    };
+
+                    fileReader.onerror = function (e) {
+                        error(e.target.error);
```
Comment (jseremba) [inline src/CameraWidgetForPhoneGap/widget/CameraWidgetForPhoneGap.js]: Same here `event`

File: src/CameraWidgetForPhoneGap/CameraWidgetForPhoneGap.webmodeler.js [hunk 13]
```diff
@@ -18,16 +18,16 @@ module.exports.preview = React.createClass({
...
+        var button = React.DOM.button({ className: "btn btn-primary" + this.widgetClass + "-button" + this.props.buttonClass + this.widgetClass + "-" + this.alignment }, this.props.buttonText);
         var elements = [ divPreview, button ];
 
         if (/below|right/.test(this.props.imageLocation)) {
             elements.reverse();
         }
 
         return React.DOM.div(
-            { className: "widget-google-maps" },
+            { className: "widget-camera-phonegap" },
```
Comment (jseremba) [inline src/CameraWidgetForPhoneGap/CameraWidgetForPhoneGap.webmodeler.js:30]: Class name does not match any classes in the widget `css` file there exists `wx-CameraWidgetForPhoneGap`

File: src/CameraWidgetForPhoneGap/widget/CameraWidgetForPhoneGap.js (hunk 11 above)
Comment (McDoyen) [reply src/CameraWidgetForPhoneGap/widget/CameraWidgetForPhoneGap.js]: Done

File: package.json (hunk 5 above)
Comment (jseremba) [reply package.json]: Done

File: package.json (hunk 6 above)
Comment (jseremba) [reply package.json]: Done

File: package.json (hunk 7 above)
Comment (jseremba) [reply package.json]: Re-Added

File: src/CameraWidgetForPhoneGap/CameraWidgetForPhoneGap.xml (hunk 8 above)
Comment (jseremba) [reply src/CameraWidgetForPhoneGap/CameraWidgetForPhoneGap.xml]: Done

File: src/CameraWidgetForPhoneGap/CameraWidgetForPhoneGap.xml (hunk 9 above)
Comment (jseremba) [reply src/CameraWidgetForPhoneGap/CameraWidgetForPhoneGap.xml]: Renamed

File: src/CameraWidgetForPhoneGap/widget/CameraWidgetForPhoneGap.js (hunk 10 above)
Comment (jseremba) [reply src/CameraWidgetForPhoneGap/widget/CameraWidgetForPhoneGap.js]: Message improved
